File: DPipeSourceCPP/impl/dpserver.cpp
#include "dpserver.h"

using namespace std;
using namespace crdk::dpipes;

//DPReceivedRequest implementation
#pragma region DPSendingResponse
	DPSendingResponse::DPSendingResponse(GUID guid) {
		_guid = guid;
	}

	GUID DPSendingResponse::guid() const {
		return _guid;
	}
#pragma endregion DPSendingResponse

//DPReceivedRequest implementation
#pragma region DPReceivedRequest
	DPReceivedRequest::DPReceivedRequest(
		GUID guid,
		int code,
		int descriptorType,
		DWORD descriptorSize,
		bool descriptorAllocated,
		std::shared_ptr<HeapAllocatedData> descriptor,
		int dataType,
		DWORD dataSize,
		std::shared_ptr<HeapAllocatedData> data,
		IDPServer* server) {
		_guid = guid;
		_code = code;
		_descriptorType = descriptorType;
		_descriptorSize = descriptorSize;
		_descriptorAllocated = descriptorAllocated;
		_descriptor = descriptor;
		_dataType = dataType;
		_dataSize = dataSize;
		_data = data;
		_server = server;
	}

	DPReceivedRequest::~DPReceivedRequest() { }

	GUID DPReceivedRequest::guid() const {
		return _guid;
	}

	int DPReceivedRequest::code() const {
		return _code;
	}

	bool DPReceivedRequest::descriptorAllocated() const {
		return _descriptorAllocated;
	}

	int DPReceivedRequest::descriptorType() const {
		return _descriptorType;
	}

	DWORD DPReceivedRequest::descriptorSize() const {
		return _descriptorSize;
	}

	std::shared_ptr<HeapAllocatedData> DPReceivedRequest::descriptor() {
		return _descriptor;
	}

	void* DPReceivedRequest::descriptorPtr() const {
		return _descriptor->data();
	}

	bool DPReceivedRequest::dataAllocated() const {
		return _dataAllocated;
	}

	int DPReceivedRequest::dataType() const {
		return _dataType;
	}

	DWORD DPReceivedRequest::dataSize() const {
		return _dataSize;
	}

	shared_ptr<HeapAllocatedData> DPReceivedRequest::data() {
		return _data;
	}

	void* DPReceivedRequest::dataPtr() const {
		return _data->data();
	}

	bool DPReceivedRequest::keepDescriptor() {
		return _descriptor->keepAllocatedMemory;
	}

	void DPReceivedRequest::keepDescriptor(bool keepAllocatedMemory) {
		_descriptor->keepAllocatedMemory = keepAllocatedMemory;
	}

	bool DPReceivedRequest::keepData() {
		return _data->keepAllocatedMemory;
	}

	void DPReceivedRequest::keepData(bool keepAllocatedMemory) {
		_data->keepAllocatedMemory = keepAllocatedMemory;
	}

	void DPReceivedRequest::ReadDescriptor(bool keepAllocatedMemory) {

		if (!_descriptorAllocated && _descriptorSize > 0) {

			_descriptor.~shared_ptr();

			_descriptor = make_shared<HeapAllocatedData>(_descriptorSize, keepAllocatedMemory);
			_server->Pipe()->Read(_descriptor->data(), _descriptorSize);
			_descriptorAllocated = true;
		}

	}
	void DPReceivedRequest::Read(bool keepAllocatedMemory) {

		ReadDescriptor(keepAllocatedMemory);

		if (!_dataAllocated && _dataSize > 0) {

			_data.~shared_ptr();

			_data = make_shared<HeapAllocatedData>(_dataSize, keepAllocatedMemory);
			_server->Pipe()->Read(_data->data(), _dataSize);
			_dataAllocated = true;
		}
	}

	DPSendingResponse DPReceivedRequest::createRespose() {
		return DPSendingResponse(guid());
	}
	IDPServer* DPReceivedRequest::server() const {
		return _server;
	}

#pragma endregion DPReceivedRequest

//DPServer implementation
#pragma region DPServer

	DPServer::DPServer(IDPipe* dpipe, bool handleAsync) :
		IDPipeMessanger(dpipe)
	{
		if (_dpipe == nullptr)
			return;

		_handleRequestAsync = handleAsync;

		_nBufferRequestSize = DP_REQUEST_SIZE;
		_pBufferRequest = new char[DP_REQUEST_SIZE];

		_nBufferResponseSize = DP_RESPONSE_SIZE;
		_pBufferResponse = new char[DP_RESPONSE_SIZE];

		_dpipe->SetClientConnectCallback([=](IDPipe* pipe, PacketHeader header) { OnClientConnect(pipe, header); });
		_dpipe->SetOtherSideDisconnectCallback([=](IDPipe* pipe, PacketHeader header) { OnClientDisconnect(pipe, header); });
		_dpipe->SetPacketHeaderRecevicedCallback([=](IDPipe* pipe, PacketHeader header) { OnDataReceived(pipe, header); });
	}

	DPServer::~DPServer() { 
		_dpipe->SetClientConnectCallback({});
		_dpipe->SetOtherSideDisconnectCallback({});
		_dpipe->SetPacketHeaderRecevicedCallback({});

		delete[] _pBufferRequest;
		delete[] _pBufferResponse;
	}

	void DPServer::OnClientConnect(IDPipe* pipe, PacketHeader header) {
		if (_onClientConnect)
			_onClientConnect(pipe, header);
	}

	void DPServer::OnClientDisconnect(IDPipe* pipe, PacketHeader header) {
		if (_onClientDisconnect)
			_onClientDisconnect(pipe, header);
	}

	void DPServer::OnRequestReceived(IDPipe* pipe, DPRequestHeader& reqRecord, DPReceivedRequest req) {

		bool handlerFound = false;

		function<void(DPReceivedRequest&)> handlerFunction = nullptr;

		_handlerMutex.lock();
			for (DPHandler const &handler : _handlers) {
				if (handler.code == reqRecord.code) {
					handlerFound = true;
					handlerFunction = handler.function;
				}
			}
		_handlerMutex.unlock();

		if (handlerFunction != nullptr)
			handlerFunction(req);

		if (!handlerFound) {
			auto resp = req.createRespose();
			resp.code = DP_HANDLER_NOT_FOUND;
			req.keepDescriptor(false);
			req.keepData(false);
			SendResponse(req, resp);
		}
	}

	bool DPServer::keepDescriptorMemoryAllocated() {
		return _keepDescriptorMemoryAllocated;
	}

	void DPServer::keepDescriptorMemoryAllocated(bool value) {
		_keepDescriptorMemoryAllocated = value;
	}

	bool DPServer::keepRequestMemoryAllocated() {
		return _keepRequestMemoryAllocated;
	}

	void DPServer::keepRequestMemoryAllocated(bool value) {
		_keepRequestMemoryAllocated = value;
	}

	IDPipe* DPServer::Pipe() {
		return _dpipe;
	}

	DWORD DPServer::maxDescriptorMemoryAllocation() {
		return _maxDescriptorMemoryAllocation;
	}

	void DPServer::maxDescriptorMemoryAllocation(DWORD value) {
		_maxDescriptorMemoryAllocation = value;
	}

	void DPServer::maxRequestMemoryAllocation(DWORD value) {
		_maxRequestMemoryAllocation = value;
	}

	void DPServer::maxRequestMemoryAllocationMB(DWORD value) {
		_maxRequestMemoryAllocation = value * 1048576;
	}

	DWORD DPServer::maxRequestMemoryAllocation() {
		return _maxRequestMemoryAllocation;
	}

	DWORD DPServer::maxRequestMemoryAllocationMB() {
		return _maxRequestMemoryAllocation / 1048576;
	}

	void DPServer::OnRequestReceivedAsync(LPVOID dataPtr) {
		auto requestData = (DPRequestReceivedAsyncData*)dataPtr;
		requestData->server->OnRequestReceived(requestData->pipe, requestData->reqRecord, requestData->request);
		delete requestData;
	}

	void DPServer::OnDataReceived(IDPipe* pipe, PacketHeader header) {
		
		DWORD nBytesRead;
		//Get data code ignoring first 8 bits (using for encoding)
		DWORD dataCode = header.GetDataCodeOnly();

		//ReadData from Data Request
		if (dataCode == DP_REQUEST) {

			DWORD dataSize = header.DataSize();

			if (dataSize < DP_REQUEST_SIZE)
				throw exception("Received broken request");

			_dpipe->Read(_pBufferRequest, DP_REQUEST_SIZE, &nBytesRead, NULL);
			DPRequestHeader reqRecord;
			FillRequestRecord(reqRecord);

			bool descriptorAllocated;
			DWORD descriptorSize = reqRecord.descriptorSize;
			shared_ptr<HeapAllocatedData> descriptorData = nullptr;

			if (descriptorSize <= _maxDescriptorMemoryAllocation) {
				descriptorData = make_shared<HeapAllocatedData>(descriptorSize, _keepDescriptorMemoryAllocated);
				DWORD nBytesRead;
				_dpipe->Read(descriptorData->data(), descriptorSize, &nBytesRead);
				descriptorAllocated = true;
			}
			else {
				descriptorData = make_shared<HeapAllocatedData>(0, _keepDescriptorMemoryAllocated);
				descriptorAllocated = false;
			}

			bool dataAllocated;
			dataSize = dataSize - DP_RESPONSE_SIZE - descriptorSize;
			shared_ptr<HeapAllocatedData> data = nullptr;

			if (descriptorAllocated && dataSize <= _maxRequestMemoryAllocation) {
				data = make_shared<HeapAllocatedData>(dataSize, _keepRequestMemoryAllocated);
				DWORD nBytesRead;
				_dpipe->Read(data->data(), dataSize, &nBytesRead);
				dataAllocated = true;
			}
			else {
				data = make_shared<HeapAllocatedData>(0, _keepRequestMemoryAllocated);
				dataAllocated = false;
			}

			DPReceivedRequest req(
				reqRecord.guid,
				reqRecord.code,
				reqRecord.descriptorType,
				reqRecord.descriptorSize,
				descriptorAllocated,
				descriptorData,
				reqRecord.dataType,
				dataSize,
				data,
				this
			);

			if (_handleRequestAsync) {
				DWORD _nReadThreadId;
				DPRequestReceivedAsyncData* reqData = new DPRequestReceivedAsyncData(this, reqRecord, req, _dpipe);
				auto _tHandleThread = CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)OnRequestReceivedAsync, (LPVOID)reqData, 0, &_nReadThreadId);
			}
			else
				OnRequestReceived(pipe, reqRecord, req);

			return;
		}

		bool isStringData = dataCode & 0x01;

		//If data is string
		if (isStringData) {

			DWORD dataSize = header.DataSize();
			auto heapData = make_shared<HeapAllocatedData>(dataSize, false);

			if (dataSize > 0) {
				void* buffer = heapData.get()->data();
				_dpipe->Read(buffer, header.DataSize(), &nBytesRead, NULL);
			}

			if (_handleRequestAsync) {
				DPMessageReceivedAsyncData* messageData = new DPMessageReceivedAsyncData(this, header, heapData, pipe);

				DWORD _nReadThreadId;
				auto _tHandleThread = CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)OnMessageStringReceivedAsync, (LPVOID)messageData, 0, &_nReadThreadId);
			}
			else
				OnMessageStringReceived(pipe, header, heapData);
		}

		//If data is binary
		else {
			DWORD dataSize = header.DataSize();
			auto heapData = make_shared<HeapAllocatedData>(dataSize, false);

			if (dataSize > 0) {
				void* buffer = heapData.get()->data();
				_dpipe->Read(buffer, header.DataSize(), &nBytesRead, NULL);
			}

			if (_handleRequestAsync) {
				DPMessageReceivedAsyncData* messageData = new DPMessageReceivedAsyncData(this, header, heapData, pipe);

				DWORD _nReadThreadId;
				auto _tHandleThread = CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)OnMessageReceivedAsync, (LPVOID)messageData, 0, &_nReadThreadId);
			}
			else
				OnMessageReceived(pipe, header, heapData);
		}
	}

	void DPServer::FillRequestRecord(const DPRequestHeader& reqRecord) const {
		memcpy_s((char*)&reqRecord.guid, sizeof(GUID), _pBufferRequest, sizeof(GUID));
		memcpy_s((char*)&reqRecord.code, sizeof(int), _pBufferRequest + sizeof(GUID), sizeof(int));
		memcpy_s((char*)&reqRecord.dataType, sizeof(int), _pBufferRequest + sizeof(GUID) + sizeof(int), sizeof(int));
		memcpy_s((char*)&reqRecord.descriptorType, sizeof(int), _pBufferRequest + sizeof(GUID) + sizeof(int) * 2, sizeof(int));
		memcpy_s((char*)&reqRecord.descriptorSize, sizeof(int), _pBufferRequest + sizeof(GUID) + sizeof(int) * 3, sizeof(int));
	}

	void DPServer::PrepareResponce(const DPSendingResponse& response, const GUID& guid) {
		memcpy_s(_pBufferResponse, sizeof(GUID), (char*)&guid, sizeof(GUID));
		memcpy_s(_pBufferResponse + sizeof(GUID), sizeof(int), (char*)&response.code, sizeof(int));
		memcpy_s(_pBufferResponse + sizeof(GUID) + sizeof(int), sizeof(int), (char*)&response.dataType, sizeof(int));
		memcpy_s(_pBufferResponse + sizeof(GUID) + sizeof(int) * 2, sizeof(int), (char*)&response.descriptorType, sizeof(int));
		memcpy_s(_pBufferResponse + sizeof(GUID) + sizeof(int) * 3, sizeof(int), (char*)&response.descriptorSize, sizeof(int));
	}

	void DPServer::SendResponse(const DPReceivedRequest& req, const DPSendingResponse& resp)
	{
		_mutexWrite.lock();
			PrepareResponce(resp, req.guid());
			DWORD nBytesWritten;
			DWORD dataLen = _nBufferResponseSize + resp.dataSize;
			PacketHeader header(false, dataLen);
			header.SetDataCode(DP_RESPONSE);

			_dpipe->WritePacketHeader(header);
			_dpipe->WriteRaw(_pBufferResponse, DP_RESPONSE_SIZE, &nBytesWritten);

			if (resp.dataSize > 0) {
				if (resp.data != nullptr)
					_dpipe->WriteRaw(resp.data, resp.dataSize, &nBytesWritten);
				else
					throw exception("Unable to send response data: buffer is null");
			}
		_mutexWrite.unlock();
	}

	void DPServer::SetHandler(int code, function<void(DPReceivedRequest&)> function) {
		DPHandler handler;
		handler.code = code;
		handler.function = function;

		_handlerMutex.lock();
			_handlers.push_back(handler);
		_handlerMutex.unlock();
	}

	void DPServer::RemoveHandler(int code) {
		_handlerMutex.lock();
		for (auto it = _handlers.begin(); it != _handlers.end(); it++) {
			if (it->code == code) {
				it->function = nullptr;
				_handlers.erase(it);
				break;
			}
		}
		_handlerMutex.unlock();
	}

	void DPServer::SetOnClientConnectCallback(function<void(IDPipe* pipe, PacketHeader header)> function) {
		_onClientConnect = function;
	}
	
	void DPServer::SetOnClientDisconnectCallback(function<void(IDPipe* pipe, PacketHeader header)> function) {
		_onClientDisconnect = function;
	}

	void DPServer::FreeAllocatedMemory(void* data) {
		HeapAllocatedData::Free(data);
	}

	void DPServer::FreeAllocatedMemory() {
		HeapAllocatedData::Free();
	}

	void DPServer::Disconnect() {
		_dpipe->Disconnect();
	}

	void DPServer::Disconnect(LPCVOID pConnectData, DWORD nConnectDataSize) {
		_dpipe->Disconnect(pConnectData, nConnectDataSize);
	}

	bool DPServer::Start() {
		return _dpipe->Start();
	}

	shared_ptr<IDPipeHandle> DPServer::GetHandle() {
		return _dpipe->GetHandle();
	}
#pragma endregion DPServer

	DPRequestReceivedAsyncData::DPRequestReceivedAsyncData(
		IDPServer* serverPtr,
		DPRequestHeader reqRec, 
		DPReceivedRequest clientRequest, 
		IDPipe* pipePtr) :
		request(clientRequest)
	{
		server = serverPtr;
		reqRecord = reqRec;
		pipe = pipePtr;
	}

File: CreatorDK.IO.DPipes/DPServer/DPServer.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace CreatorDK.IO.DPipes
{
    public interface IDPServer
    {
        DPipe Pipe { get; }

        void SendResponse(DPReceivedRequest request, DPClientResponse response);
    }

    public class DPServer : DPMessangerBase, IDPServer
    {
        public DPServer(DPipe dpipe, bool handleAsync, Encoding encoding = null, uint stringBufferSize = 4096) : 
            base(dpipe, encoding, stringBufferSize)
        {
            _handleAsync = handleAsync;

            _dpipe.OnOtherSideConnectCallback = OnClientConnectInner;
            _dpipe.OnOtherSideDisconnectCallback = OnClientDisonnectInner;
            _dpipe.OnPacketHeaderReceivedCallback = OnDataReceived;
        }

        ~DPServer()
        {
            _dpipe.OnOtherSideConnectCallback = null;
            _dpipe.OnOtherSideDisconnectCallback = null;
            _dpipe.OnPacketHeaderReceivedCallback = null;
        }

        private bool _handleAsync;
        private byte[] _pBufferRequest = new byte[Constants.DP_REQUEST_SIZE];
        private byte[] _pBufferResponse = new byte[Constants.DP_RESPONSE_SIZE];

        private Dictionary<int, DPHandler> _handlers = new Dictionary<int, DPHandler>();
        private Mutex _handlerMutex = new Mutex(false);

        public Action<PacketHeader> OnClientConnect;
        public Action<PacketHeader> OnClientDisconnect;

        private int _maxDescriptorMemoryAllocation = 33554432;  //32 MB
        private int _maxRequestMemoryAllocation = 33554432;     //32 MB

        public int MaxDescriptorMemoryAllocation
        {
            get => _maxDescriptorMemoryAllocation;
            set => _maxDescriptorMemoryAllocation = value;
        }
        public int MaxDescriptorMemoryAllocationMB
        {
            get => _maxDescriptorMemoryAllocation / 1048576;
            set => _maxDescriptorMemoryAllocation = value * 1048576;
        }

        public int MaxRequestMemoryAllocation
        {
            get => _maxRequestMemoryAllocation;
            set => _maxRequestMemoryAllocation = value;
        }
        public int MaxRequestMemoryAllocationMB
        {
            get => _maxRequestMemoryAllocation / 1048576;
            set => _maxRequestMemoryAllocation = value * 1048576;
        }

        public DPipe Pipe => _dpipe;

        private void OnClientConnectInner(PacketHeader header)
        {
            OnClientConnect?.Invoke(header);
        }
        private void OnClientDisonnectInner(PacketHeader header)
        {
            OnClientDisconnect?.Invoke(header);
        }
        private void OnDataReceived(PacketHeader header)
        {
            var command = header.DataCode;

            //ReadData from Data Request
            if (command == Constants.DP_REQUEST)
            {
                if (header.DataSize < Constants.DP_REQUEST_SIZE)
                    throw new Exception("Received broken request");

                _dpipe.Read(_pBufferRequest, 0, (int)Constants.DP_REQUEST_SIZE);
                var requestHeader = GetRequestHeader();

                bool descriptorAllocated;
                int descriptorSize = requestHeader.DescriptorSize;
                byte[] descriptor = null;

                if (descriptorSize <= _maxDescriptorMemoryAllocation)
                {
                    descriptor = new byte[descriptorSize];
                    _dpipe.Read(descriptor, 0, descriptorSize);
                    descriptorAllocated = true;
                }
                else
                    descriptorAllocated = false;

                bool dataAllocated;
                int dataSize = header.DataSize - (int)Constants.DP_RESPONSE_SIZE - descriptorSize;
                byte[] data = null;

                if (descriptorAllocated && dataSize <= _maxRequestMemoryAllocation)
                {
                    data = new byte[dataSize];
                    _dpipe.Read(data, 0, dataSize);
                    dataAllocated = true;
                }
                else
                    dataAllocated = false;


                //DPClientRequest request = new DPClientRequest(requestHeader.Guid, requestHeader.Code, requestHeader.DataType, data);
                DPReceivedRequest request = new DPReceivedRequest(
                    requestHeader.Guid,
                    requestHeader.Code,
                    requestHeader.DescriptorType,
                    requestHeader.DescriptorSize,
                    descriptorAllocated,
                    descriptor,
                    requestHeader.DataType,
                    dataSize,
                    dataAllocated,
                    data,
                    this
                    );


                if (_handleAsync)
                    Task.Run(() => OnRequestReceived(requestHeader, request));
                else
                    OnRequestReceived(requestHeader, request);

                return;
            }

            bool isStringData = (command & 0x01) > 0;

            //If data is string
            if (isStringData)
            {
                string message = GetString(header);

                if (_handleAsync)
                {
                    Task.Run(() => OnMessageStringReceivedInner(header, message));
                }
                else
                {
                    OnMessageStringReceivedInner(header, message);
                }
            }
            //If data is binary
            else
            {
                byte[] messageData = _dpipe.Read(header);

                if (_handleAsync)
                {
                    Task.Run(() => OnMessageDataReceivedInner(header, messageData));
                }
                else
                {
                    OnMessageDataReceivedInner(header, messageData);
                }
            }
        }

        private void OnRequestReceived(DPRequestHeader requestHeader, DPReceivedRequest request)
        {
            bool handlerExists = false;

            DPHandler functionDelegate = null;

            _handlerMutex.WaitOne();
            if (_handlers.ContainsKey(requestHeader.Code))
            {
                handlerExists = true;
                functionDelegate = _handlers[requestHeader.Code];
            }
            _handlerMutex.ReleaseMutex();

            if (functionDelegate != null)
                functionDelegate.Invoke(request);

            if (!handlerExists)
            {
                var respopnse = request.CreateResponse();
                respopnse.Code = (int)Constants.DP_HANDLER_NOT_FOUND;
                SendResponse(request, respopnse);
            }
        }
        private void PrepareResponce(DPClientResponse response, Guid guid)
        {
            byte[] guidBytes = guid.ToByteArray();
            byte[] codeBytes = BitConverter.GetBytes(response.Code);
            byte[] dataTypeBytes = BitConverter.GetBytes(response.DataType);
            byte[] descriptinTypeBytes = BitConverter.GetBytes(response.DescriptionType);
            byte[] descriptinSizeBytes = BitConverter.GetBytes(response.DescriptionSize);

            Buffer.BlockCopy(guidBytes, 0, _pBufferResponse, 0, guidBytes.Length);
            Buffer.BlockCopy(codeBytes, 0, _pBufferResponse, 16, codeBytes.Length);
            Buffer.BlockCopy(dataTypeBytes, 0, _pBufferResponse, 20, dataTypeBytes.Length);
            Buffer.BlockCopy(descriptinTypeBytes, 0, _pBufferResponse, 24, descriptinTypeBytes.Length);
            Buffer.BlockCopy(descriptinSizeBytes, 0, _pBufferResponse, 28, descriptinSizeBytes.Length);
        }
        private DPRequestHeader GetRequestHeader()
        {
            byte[] guidBytes = new byte[16];
            Buffer.BlockCopy(_pBufferRequest, 0, guidBytes, 0, 16);

            return new DPRequestHeader()
            {
                //Guid = new Guid(_pBufferRequest),
                Guid = new Guid(guidBytes),
                Code = BitConverter.ToInt32(_pBufferRequest, 16),
                DataType = BitConverter.ToInt32(_pBufferRequest, 20),
                DescriptorType = BitConverter.ToInt32(_pBufferRequest, 24),
                DescriptorSize = BitConverter.ToInt32(_pBufferRequest, 28),
            };
        }
        public void SendResponse(DPReceivedRequest request, DPClientResponse response)
        {
            _mutexWritePipe.WaitOne();

            PrepareResponce(response, request.Guid);
            int dataLen = _pBufferResponse.Length + (response.Data == null ? 0 : response.Data.Length);
            PacketHeader packetHeader = new PacketHeader(dataLen, Constants.DP_RESPONSE);
            _dpipe.WritePacketHeader(packetHeader);
            _dpipe.WriteRaw(_pBufferResponse, 0, (int)Constants.DP_RESPONSE_SIZE);

            if (response.Data != null && response.Data.Length > 0) 
                _dpipe.WriteRaw(response.Data, 0, response.Data.Length);

            _mutexWritePipe.ReleaseMutex();
        }
        public void SetHandler(int code, DPHandler handler)
        {
            _handlerMutex.WaitOne();
                _handlers.Add(code, handler);
            _handlerMutex.ReleaseMutex();
        }
        public void RemoveHandler(int code)
        {
            _handlerMutex.WaitOne();
            _handlers.Remove(code);
            _handlerMutex.ReleaseMutex();
        }
        public void Start()
        {
            _dpipe.Start();
        }
        public void Disconnect(byte[] data = null)
        {
            _dpipe.Disconnect(data);
        }
    }
}
